optuna_dashboard/ts/components/GraphParetoFront.tsx

```diff
@@ -165,6 +165,58 @@ const makeMarker = (
   }
 }
 
+const getIsDominatedTrialND = (normalizedValues: number[][]) => {
+  // Fallback for straight-forward pareto front algorithm (O(N^2) complexity).
+  const dominatedTrials: boolean[] = []
+  normalizedValues.forEach((values0: number[]) => {
+    const dominated = normalizedValues.some((values1: number[]) => {
+      if (values0.every((value0: number, k: number) => values1[k] === value0)) {
+        return false
+      }
+      return values0.every((value0: number, k: number) => values1[k] <= value0)
+    })
+    dominatedTrials.push(dominated)
+  })
+  return dominatedTrials
+}
+
+const getIsDominatedTrial2D = (normalizedValues: number[][]) => {
+  // Fast pareto front algorithm (O(N log N) complexity).
+  const sorted = normalizedValues
+    .map((values, i) => [values[0], values[1], i])
+    .sort()
```

Comment:
```ts
const values = [[2, 2], [3, 2], [10, 2]]
const sortedValues = values.map((v, i) => [v[0], v[1], i]).sort()
console.log(sortedValues)
```

Output

```
[ [ 10, 2, 2 ], [ 2, 2, 0 ], [ 3, 2, 1 ] ]
```

So you still need to change the comparator inside `sort()`.

```suggestion
    .sort((a, b) => {
        if (a[0] !== b[0]){
            return a[0] < b[0]
        }else if (a[1] !== b[1]){
            return a[1] < b[1]
        }else{
            return a[2] < b[2]
        }
    })
```